src/storage/table/row_group_collection.cpp

```diff
@@ -656,32 +656,59 @@ void RowGroupCollection::Update(TransactionData transaction, row_t *ids, const v
 void RowGroupCollection::RemoveFromIndexes(TableIndexList &indexes, Vector &row_identifiers, idx_t count) {
 	auto row_ids = FlatVector::GetData<row_t>(row_identifiers);
 
-	// initialize the fetch state
-	// FIXME: we do not need to fetch all columns, only the columns required by the indices!
+	// collect all indexed columns
+	unordered_set<column_t> indexed_column_id_set;
+	indexes.Scan([&](Index &index) {
+		auto set = index.GetColumnIdSet();
+		indexed_column_id_set.insert(set.begin(), set.end());
+		return false;
+	});
+
+	// initialize the fetch state with only indexed columns
 	TableScanState state;
+
 	vector<StorageIndex> column_ids;
-	column_ids.reserve(types.size());
-	for (idx_t i = 0; i < types.size(); i++) {
-		column_ids.emplace_back(i);
+	column_ids.reserve(indexed_column_id_set.size());
+	for (auto &col : indexed_column_id_set) {
+		column_ids.emplace_back(col);
+	}
+	sort(column_ids.begin(), column_ids.end());
+
+	vector<LogicalType> column_types;
+	column_types.reserve(indexed_column_id_set.size());
+	for (auto &col : column_ids) {
+		column_types.push_back(types[col.GetPrimaryIndex()]);
 	}
+
 	state.Initialize(std::move(column_ids));
 	state.table_state.max_row = row_start + total_rows;
 
-	// initialize the fetch chunk
+	// create two chunks:
+	// result: used for scanning data, has only the required columns
+	// rows: used for index removal, has all columns but only initializes indexed ones
 	DataChunk result;
-	result.Initialize(GetAllocator(), types);
+	result.Initialize(GetAllocator(), column_types);
```

Comment:
Nit: The names `result` and `rows` are confusing. Can we use something like this? To distinguish between their purposes.
```cpp
// fetched_columns_chunk
// all_columns_chunk
```